### ISSUE TITLE
Remove redundancy in table's `cols`

### ISSUE THREAD

**Issue by lukasjuhrich** [agdsn/pycroft#52]
_(this is mainly a self-reminder and partially unordered thoughts.)_

The columns for a table are defined in a redundant manner – one in the html template, and once in the view function providing the json endpoint, since it provides a json dict with the columns as keys. At least one should not have to change colums in the html template when changing the json view, but in the view function and the json view. This logic doesn't belong in the template imho.

For instance, compare https://github.com/agdsn/pycroft/blob/develop/web/blueprints/user/__init__.py#L220 to https://github.com/agdsn/pycroft/blob/develop/web/templates/user/_user_show_hosts.html#L7
## EDIT:

See [this gist](https://gist.github.com/lukasjuhrich/28d976c44fd38bfc5ca407f27632c192).

---
## First ideas

One could solve it with providing the colums in an attribute given in the view function (for the page, not the table's json endpoint).  Then, it can be refactored out without complicated dependencies since both cols are in the same file.

One could spin this idea further and provide an object (most likely `namedtuple(…)()`) that provides all the necessary information and might carry the endpoint's name as well. Our template would reduce to

``` python
{% import "macros/table.html" as tbl %}
{{ tbl.table_from_object(host_table_info) }}
{# dir(host_table_info) contains {'columns': ['foo', 'bar'], 'json_endpoint': 'user.user_show_hosts_json'} #}
```

where `table_from_object` is a macro using `item_table` with the information from `host_table_info`.

the json view function might make use of some connected class like this:

``` python
@bp.route("/<int:user_id>/hosts")
def user_show_hosts_json(user_id):
    list_items = []
    for user_host in User.q.get(user_id).user_hosts:
        # …
        for ip in user_host.ips:
            # use the classmethod create_json_dict:
            list_items.append(UserHostsTableInfo.create_json_dict(
                id=str(user_host.id),
                ip=str(ip.address),
                mac=ip.interface.mac,
                switch=switches,
                port=ports,
                action=foo,
            ))
    return jsonify(items=list_items)
```

The redundancy is technically still there, but easier to test since it must match the function signature.

Perhaps this can be solved more elegantly still.

**Comment by lukasjuhrich** [comment]
This has been resolved conceptually by abstracting the tables to the layer of python classes. The implicit contract that the JSON Endpoint used by some table should actually provide the correct scheme (as a subset, to be accurate) still holds, but I don't think it would necessarily make things better if one would try to enforce the scheme in the JSON view definition using the table classes.

Perhaps, however, it could be useful to add a validation routine throwing warnings in each endpoint used by a `bootstrap-table`.

Although the redundancy has not been technically resolved, it is much easier to find the column definitions because one can utilize the usual python-IDE workflow of looking definitions up – in this case the definition of the table object being passed to the template. Before, one had to look into the template, find where the given JSON URL has been used, and then where the columns have been defined.

**Comment by lukasjuhrich** [comment]
I finally have a concept that could help us with that! The idea arose with changes in #484.

The main pain point is that structured data is passed as a dict, and there is no way for the IDE or python to assist you in finding the right key names – was it `href` or `url` for the `linkFormatter`? Was it named `linkFormatter` or `urlFormatter`? Was there a `type` field? etc.

This can be resolved by adding a [generic dict pass-through method](https://github.com/agdsn/bootstrap-table-py/compare/ea750fcd7eca58a4cf2668867f93e8f21475d0b3...933afd787dfa2cbe0824e6ea619ae32e18f2d8c6#diff-77610039486a5e8329752eba56879c9de7a8ae78dddd71ae15bb265458e311aaR98-R116) `value()` on a `Column`, and communicating the expected keys in the type hints of e.g. [the actual `LinkColumn`](https://github.com/agdsn/bootstrap-table-py/compare/ea750fcd7eca58a4cf2668867f93e8f21475d0b3...933afd787dfa2cbe0824e6ea619ae32e18f2d8c6#diff-77610039486a5e8329752eba56879c9de7a8ae78dddd71ae15bb265458e311aaR145-R150):

```python
@custom_formatter_column('table.linkFormatter')
class LinkColumn(DictValueMixin, Column):
    pass
    if typing.TYPE_CHECKING:
        @classmethod
        def value(cls, href: str, title: str, glyphicon: Optional[str] = None) -> dict: ...
```

Then, the value dicts can just be constructed like
```python
Table.column.value(href="https://agdsn.de", title=title)
```

A simillar thing works with [the table's row construction](https://github.com/agdsn/bootstrap-table-py/compare/ea750fcd7eca58a4cf2668867f93e8f21475d0b3...933afd787dfa2cbe0824e6ea619ae32e18f2d8c6#diff-77610039486a5e8329752eba56879c9de7a8ae78dddd71ae15bb265458e311aaR98-R116).

**Comment by lukasjuhrich** [comment]
I would suggest planning this after #485 as this requires wide-spread changes a) in the bs-table framework and b) in all the usages, and I would prefer clean commits over having to hop between submodule and actual git repo many times.

**Comment by lukasjuhrich** [comment]
Typing the `Col`s has been fixed in f6372518d83a8060bf8979de6e5898150bf21f19.

One might think about typing the row creation [as well](https://github.com/agdsn/pycroft/blob/ffb2d649fb9c31ef8e0b427b853b12a031a08ed5/web/blueprints/user/tables.py#L123..L127) in a follow-up issue.
